File: src/gowerdistance.c
/* Implement Gower distance with Podani amendment to ordered factors. */

/* Standard R headers */

#include <R.h>
#include <Rmath.h>
#include <math.h>

/* The variable type (vtype) must be one of the following
   integers. Type PRESENCE is the same as "asymmetric binary" and
   implements a Jaccard-like disismilarity. Ordered factors are
   handled either as their internal codes (METRICINTERNAL) like in
   cluster::daisy(), as their ranks (METRICORDINAL) or as ranks with
   Podani's tie correction. Normal quantitative variables are
   METRIC. TIE is a special case: the input matrix must be amended
   with columns which give number of tied values for each observation
   of each ORDINAL variable.
*/

#define BINARY 1
#define PRESENCE 2
#define NOMINAL 3
#define ORDINAL 4
#define METRICORDINAL 5
#define METRICINTERNAL 6
#define METRIC 7
#define TIE 8

/* Function adds two arguments to the normal call. Both of these are
   vectors of lenght nc:

   vtype: variable type as #define'd above

   scale: variable to scale each column to equal weight in Gower
   distance. For most vtype's this is 1, but for METRIC, METRICORDINAL
   and METRICINTERNAL it is the range, and for ORDINAL and TIE it is
   the Podani denominator. These must be calculated in the calling
   program.

   NB. This does not have weights but adding these is trivial (or they
   can be incorporated in scale).

*/

double xx_gowerpodani(double *x, int nr, int nc, int i1, int i2,
		      int *vtype, double *scale)
{
    double dist;
    int j, count;

    dist = 0.0;
    count = 0;

    for (j=0; j<nc; j++ ) {
	if (R_FINITE(x[i1]) && R_FINITE(x[i2])) {
	    switch(vtype[j]) {
	    case METRICORDINAL:
	    case METRICINTERNAL:
	    case METRIC:
	    case ORDINAL:
		dist += fabs(x[i1]-x[i2])/scale[j];
		count++;
		break;
	    case BINARY:
	    case PRESENCE:
		if (x[i1] > 0 || x[i2] > 0) {
		    dist += fabs(x[i1]-x[i2])/scale[j];
		    count++;
		}
		break;
	    case NOMINAL:
		dist += (x[i1] == x[i2]) ? 0.0 : 1.0 ;
		count++;
		break;
	    case TIE:
		dist -= ((x[i1] + x[i2]) / 2.0 - 1.0) / scale[j];
		break;
	    }
	}
	i1 += nr;
	i2 += nr;
    }
    if (count == 0) return NA_REAL;
    return dist / (double) count;
}

/* A note on implementation:
   
   Handling of PRESENCE only differs in treating double zeros: these
   do not update count and hence missing species do not reduce
   dissimilarity (i.e., do not increase count in return
   dist/count). PRESENCE does not mean that data should be 0/1: they
   can quite well be quantitative and hence we divide with scale. The
   point is exactly in treating double-zeros.

   Most other variables are treated identically as Manhattan: the only
   difference comes from scale. only NOMINAL variables are handled
   with matching coefficient. If NOMINAL variables are replaced with
   their contrast matrix, they could be handled as PRESENCE.

   The handling of Podani tie correction is more opaque. It very much
   ties around defining scale and making scale correction. For other
   variables the scale is plain range, but in Podani correction the
   scale is range - (Timax-1)/2 - (Timin-1)/2 where Timax and Timin
   are number of items tied to max and min of the variable. This is a
   constant to each variable and can be supplied as scale. The
   equation 2b of Podani 1999 can be rearranged as 

   (Manhattan - Tiecorrection)/scale = 
            Manhattan/scale - Tiecorrection/scale 
   
   where

   Tiecorrection = (Tik-1)/2 + (Tjk-1)/2, 

   where T is the number of tied values for rows i and j in variable
   k. That is the column we must add to the input data. 

 */

File: R/gowerdistance.R
#' Gower Distance for Mixed Data with Tied Ranks
#'
#' Proof-of-the-concept function for implementing Gower distance
#' optionally with Podani's amendment to handling ordered factors as
#' metric data with ties.
#'
#' The function accepts a data frame of potentially different data
#' types, and used in mixed metrics of Gower distance. The data types
#' for each variable are interpreted from the data frame using
#' strategies given in \code{ordered} and \code{continuous} or as
#' given for each variable in argument \code{vtypes}.
#'
#' Factors are always interpreted as nominal variables (\code{vtypes =
#' "N"}) and analysed as Matching coefficient.
#'
#' With \code{orderred = "internal"} factor levels are replaced with
#' their internal values from 1 to number of levels (\code{vtype =
#' "O"}). Strategy \code{"metric"} (\code{vtypes = "OR"}) replaces
#' them with their averaged ranks from 1 to number of observations
#' (but due to ties usually excluding these extremes). Strategy
#' \code{ordered = "podani"} (\code{vtypes = "OP"}) does the same but
#' takes into account the ties in the analysis following Podani
#' (1999). Finally, strategy \code{"nominal"} analyses them as
#' unordered factors.
#'
#' Continuous variables can be analysed as such with strategy
#' \code{continuous = "metric"} (\code{vtypes = "C"}) using Manhattan
#' distance scaled to maximum = 1. With \code{continuous = "presence"}
#' (\code{vtypes = "P"}), double-zeros are ignored for non-negative
#' variables (but variables with negative values are treated as
#' \code{"C"}), and this gives an index that is similar to a
#' quantitative variant of Jaccard dissimilarity index. Finally, with
#' strategy \code{"binary"} double-zeros are ignored in binary
#' variables but all other numeric variables are treated as
#' \code{"C"}. This gives Jaccard index to binary variables. This case
#' is known as \dQuote{asymmetric binary} in some other
#' implementations. There is no explicit \dQuote{symmetric binary}
#' because this is similar as treating variable as \code{"C"}.

#' @return Function returns a numeric data matrix that can be used in
#' C function. Factors are replaced with their internal values,
#' ordered factors by their internal values or ranks as requested,
#' binary variables are guaranteed to have a zero level. If there are
#' any variables of \code{vtypes = "OP"}, a new column of number of
#' ties for each observation (including tie to itself) is added. Two
#' attributes are added. Attribute \code{"vtypes"} gives the
#' interpreted types as defined above and \code{vtypes = "T"} is used
#' for possible new columns of ties. Attribute \code{scale} gives the
#' scale used in C code. Usually this is the range of variables (and
#' will be ignored for \code{vtypes = "N"}), but for \code{vtypes =
#' "OP"} the scale is adjusted for the number of ties.
#'
#' 
#' @param x Input data frame possibly with different data types
#' @param ordered Policy of interpreting ordered factors (see Details).
#' @param continuous Policy of handling continuous variables (see Details).
#' @param vtypes Variable types for each column of input data. If
#' given, will take precedence over policies in \code{ordered} and
#' \code{continuous}. See Details.
#' 
#' @export
`gowerdistance` <-
    function(x, ordered = "internal", continuous = "metric", vtypes)
{
    if (!is.data.frame(x))
        x <- as.data.frame(x)
    ## allowed variable types ordered as in the C file
    typnam <- c("B","P","N","OP","OR","O","C","T")
    types <- seq_along(typnam)
    names(types) <- typnam
    ## interpret policies
    ordered <- match.arg(ordered,
                         c("internal", "metric", "podani", "nominal"))
    continuous <- match.arg(continuous,
                            c("metric", "presence", "binary"))
    ## Get variable types
    intypes <- sapply(x, function (z) class(z)[1])
    if (!missing(vtypes)) {
        if (length(vtypes) != ncol(x))
            stop("'vtypes' must contain type of every column of 'x'")
    } 
    ## interpret types
    else {
        vtypes <- rep(NA, ncol(x))
        if (any(typ <- intypes == "factor"))
            vtypes[typ] <- "N"
        if (any(typ <- intypes == "ordered"))
            vtypes[typ] <-
                switch(ordered,
                       "internal" = "O",
                       "metric" = "OR",
                       "podani" = "OP",
                       "nominal" = "N")
        ## interpretation of "numeric" can vary with data values
        if (any(typ <- intypes %in% c("numeric", "integer", "logical"))) {
            ## all are treated as metric
            if (continuous == "metric")
                vtypes[typ] <- "C"
            ## non-negative data ignore double-zeros
            if (continuous == "presence") {
                for(i in which(typ))
                    if (min(x[,i] < 0))
                        vtypes[i] <- "C"
                    else
                        vtypes[i] <- "P"
            }
            ## binary data ignore double-zeros
            if (continuous == "binary") {
                for (i in which(typ))
                    if(length(table(x[,i])) == 2)
                        vtypes[i] <- "B"
                    else
                        vtypes[i] <- "C"
            }
        }
    }
    ## start handling data
    x <- data.matrix(x)
    ## replace internal coding or ordered factors with ranks
    if (any(typ <- vtypes %in% c("OR","OP")))
        for(i in which(typ))
            x[,i] <- rank(x[,i])
    ## check that binary variables have minimum=0
    if (any(typ <- vtypes == "B"))
        for(i in which(typ))
            if(min(x[,i] != 0))
                x[,i] <- x[,i] - min(x[,i])
    ## scales for variables are usually their ranges
    scale <- apply(x, 2, function(z) diff(range(z)))
    ## but for Podani scaling we need scale adjustment
    ## -(Timin-1)/2-(Timax-1)/2 that is easiest to find from extreme
    ## averaged ranks
    if (any(typ <- vtypes == "OP"))
        for(i in which(typ))
            scale[i] <- scale[i] - min(x[,i]) + 1 - nrow(x) + max(x[,i])
    ## In Podani scaling we also need the number of tied values for
    ## each observation
    if (any(typ)) {
        ties <- apply(x[, typ, drop = FALSE], 2,
                      function(z) colSums(outer(z, z, "==")))
        x <- cbind(x, ties)
        scale <- c(scale, scale[typ])
        vtypes <- c(vtypes, rep("T", sum(typ)))
    }
    ## return x with attributes we need for Gower distance
    attr(x, "vtypes") <- types[vtypes]
    attr(x, "scale") <- scale
    x
}
